### DISCUSSION TITLE
Best way to edit content (programs, news, etc.)?

### DISCUSSION THREAD

**Discussion by jukey** [TVTower/TVTower#334] [Q&A] [answered]
In case I would like to create content in order to extend the amount of films, news, etc., what would be the best way to do so?
1. Do you use any tool that deals with the xml files?
2. Where do I find constraints like the max. amount of chars in the description of a program?

**Comment by GWRon** [comment, marked as answer]
Re 1 : No special tool available for now (always wanted to code ... but struggling with the complexity of GUI ...).

Re 2 : No real constraints as they are a bit "visually" constrained (word wrapping ...). So honestly for now it is rather "trial and error".

**Reply by GWRon** [reply]
You can - for now you "must"  - open the xml in a proper text editor (with "XML syntax highlighting" ) and raw type it in there.

**Reply by nittka** [reply]
I am currently working on an Xtext-Editor for the database files. One problem for a dedicated editor is GUI, but also serialization - you do not want to change the XML-file in places where no changes were made.
So a text-based editor has the advantage of keeping the current formatting. It will  be some time until the editor is "usable" (code completion, validation, templates for common usecases...). Also there are some inconsistencies and duplicate ways of defining the same properties in the current databse.

I'd like to do some conceptual cleaning up with Ronny. So my suggestion for the meantime - look for news/programmes that are similar to what you want to define. A very common programming approach: monkey see - monkey do.

@GWRon: Do you have to start a new game before you can see additional news/added programmes or will the database be updated even when loading a saved game (e.g. load entries that where not part of the original database)?

**Reply by GWRon** [reply]
Wir koennen uebrigens auch Deutsch schreiben - sind alle 3 Teilnehmer ja der Sprache maechtig ...


/dev loaddb 
should forcefully load a DB

@ gui editor
My code does "inline replace" of stuff without affecting the format/indentation of the XML nodes. We had one doing such a thing in a java editor too but he sadly passed away.

**Reply by nittka** [reply]
`dev loaddb` geht aber nur im Dev-Modus. In einer "finalen" Version würde das ja nicht gehen, so dass man bei neuen Programmversionen mit alten Spielständen nicht in den Genuss neuer Filme käme.
Gibt es von dem Java-Editor mit "inline-Replace" noch den Quellcode?
Wenn ich soweit bin, fände ich es schön, wenn Ihr den Xtext-Editor mal ausprobiert.

**Reply by jukey** [reply]
Danke für die Antworten. :)
Ich schaue mir die Texte dann mal mit dem 'XML Copy Editor' an. 
Wenn ich es richtig verstehe ist Xtext ein Plugin für IDEs, z.b. eclipse?

**Comment by GWRon** [comment]
Im Forum wurde das diskutiert. Muesste ich mal versuchen aufzustoebern.

**Reply by GWRon** [reply]
https://www.gamezworld.de/phpforum/viewtopic.php?id=13669

[TVTDBEdit.zip](https://github.com/TVTower/TVTower/files/6270785/TVTDBEdit.zip)

Die Jar liesse sich dekompilieren...

**Reply by nittka** [reply]
Ich habe mal kurz reingeschaut (nicht in den Code). Unterstützt werden aktuell nur Personen (Tabellenansicht mit Formular für das Editieren). Formatänderungen bei der Persistenz habe ich noch nicht geprüft.

**Comment by GWRon** [comment]
Man koennte zwecks "loaddb" eine funktion schreiben, die nur "anhaengt" statt zu ueberschreiben.

**Comment by jukey** [comment]
Eine Zusatzfrage hat sich noch ergeben (Soll ich dafürlieber eine neue Diskussion eröffnen?):
Ich mag die humorvoll an die Original-Titel angelehnten Sendungstitel. Um das hinzufügen neuer Inhalte zu erleichtern, wäre es toll, wenn im 'prorgamme' xml das Schema um den Original-Titel zu erweitern. Gleiches wäre auch mit Personen, etc. wünscenswert.

Dadurch könnte man z.B. einen Film suchen, bevor man sich einen lustigen TVT Titel für ihn ersinnt. Derzeit müsste man wirklich alle Titel einzeln durchgehen um auszuschließen, dass der titel bereits transformiert wurde.

**Reply by jukey** [reply]
Habe gerade selbst entdeckt, dass es diese Möglichkeit für 'programme' schon gibt, `<title_original>` sei dank. :)

**Comment by GWRon** [comment]
@nittka 
Kannst Du fuer den Editor-Teil eine kleine readme schreiben? Also einfach nur 2-3 Schritte, wie wir den nutzen koennen? (also Eclipse runterladen, Dort und dort klicken und die und die Datei oeffnen).

**Reply by nittka** [reply]
Ja, das mache ich im TVTDatabaseEditor-Repo. Die weitere Diskussion zu dem Thema können wir dann dort weiterführen.

**Comment by nittka** [comment]
Aktueller Stand:
* Datenbankformat ist dokumentiert
* Nutzbare Version des Datenbankeditors ist erstellt und dokumentiert (Installationsanleitung+Feature-Übersicht)
* Englische Version der Editor-Doku steht noch aus

https://github.com/TVTower/Documentation
